src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java

```diff
@@ -1,85 +1,110 @@
 package frc.robot.subsystems.ball.feeder;
 
 import ca.team3161.lib.robot.subsystem.RepeatingPooledSubsystem;
-import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
 import edu.wpi.first.wpilibj.Ultrasonic;
 import frc.robot.subsystems.ball.feeder.UltrasonicPoller;
 import ca.team3161.lib.utils.SmartDashboardTuner;
 import java.util.concurrent.TimeUnit;
+import frc.robot.subsystems.ball.feeder.conveyor.Conveyor;
+import frc.robot.subsystems.ball.feeder.conveyor.ConveyorImpl;
+import frc.robot.subsystems.ball.feeder.hopper.Hopper;
+import frc.robot.subsystems.ball.feeder.hopper.HopperImpl;
 import frc.robot.RobotMap;
 
 public class FeederImpl extends RepeatingPooledSubsystem implements Feeder {
     
-    WPI_TalonSRX beltController;
-    WPI_TalonSRX hopperController;
+    Conveyor conveyorSubsystem;
+    Hopper hopperSubsystem;
 
-    Ultrasonic feederUltrasonic;
-    UltrasonicPoller poller;
-    SmartDashboardTuner distanceTuner;
+    double topUltrasonicDistanceMM;
+    Ultrasonic topUltrasonic;
+    SmartDashboardTuner topUltrasonicTuner;
+
+    UltrasonicPoller topPoller;
 
     public FeederImpl(){
         super(1, TimeUnit.SECONDS);
         
-        this.beltController = new WPI_TalonSRX(RobotMap.BELT_TALON_PORT);
-        this.hopperController = new WPI_TalonSRX(RobotMap.HOPPER_TALON_PORT);
-        this.feederUltrasonic = new Ultrasonic(RobotMap.ULTRASONIC_PORTS[0], RobotMap.ULTRASONIC_PORTS[1]);
+        this.topUltrasonic = new Ultrasonic(RobotMap.ULTRASONIC_PORTS[0], RobotMap.ULTRASONIC_PORTS[1]);
         
+        this.conveyorSubsystem = new ConveyorImpl();
+        this.hopperSubsystem = new HopperImpl();
 
-        this.poller = new UltrasonicPoller(feederUltrasonic, 3000, 300);
-        this.distanceTuner = new SmartDashboardTuner("Ball Distance", 300, d -> this.poller.setDistance(d));
-        
-        this.hopperController.configContinuousCurrentLimit(20, 250);
-        this.hopperController.enableCurrentLimit(false);
-    }
+        this.topUltrasonicDistanceMM = 30;
+        this.topUltrasonicTuner = new SmartDashboardTuner("Top Ultrasonic Tuner", this.topUltrasonicDistanceMM, x -> this.topUltrasonicDistanceMM = x);
 
-    public void enableConveyor() {
-        this.beltController.set(-0.95d);
-        // TODO Find actual values
-    }
+        this.topPoller = new UltrasonicPoller(this.topUltrasonic, 3000, 20); // Temp values
 
-    public void enableHopper() {
-        this.hopperController.enableCurrentLimit(true);
-        this.hopperController.set(0.8d);
     }
 
-    public void stopConveyor() {
-        this.beltController.set(0.0d);
+    public void task() {
+        // PLACEHOLDER
```

Comment:
This comment is probably the "why" behind the PLACEHOLDER comment here, but it seems like most/all of the logic currently in the methods `prime`, `unload`, `enable`, and `stop` should also all be handled by setting internal state which this `task` method then attempts to reconcile over time. Right now each of those methods need to be called repeatedly externally in order to implement the correct logic - for example, `prime` being called results in the ultrasonic sensor being checked, so if the robot operator taps the button mapped to priming, then the priming operation may not actually be completed.

Reply:
Yeah, I should do that. I wasn't really thinking clearly last night and couldn't think of a way to handle that cleanly but now it's obvious.

Reply:
Do you think I should store the Hopper state and the conveyor state in separate variables? Or do you think there is a better way

Reply:
Nvm got it now